---
title: "razzo pipeline"
author: "G. Laudanno and Richel J.C. Bilderbeek"
date: "`r Sys.Date()`"
output: rmarkdown::pdf_document
vignette: >
  %\VignetteIndexEntry{razzo pipeline}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

Legenda:
nLTT = normalized lineages through time;
BD = Birth Death Model;
MBD = Multiple Birth Death Model;

This is the pipeline to follow for the Razzo project. Our aim is to compare two nLTT distributions related to two tree posteriors: one descending from MBD alignments and another one from BD alignments. Both posteriors are generated using a BD prior in BEAST.

```{r setup, results = "hide"}
if (!require(mbd)) {devtools::install_github("Giappo/mbd")}
if (!require(TESS)) {
  install.packages("TESS", repo = "https://lib.ugent.be/CRAN/")
}
if (!require(pirouette)) {
  devtools::install_github("richelbilderbeek/babette")
  devtools::install_github("richelbilderbeek/pirouette")
}
```

## Folder structure

In `folder_name`:

  * `1`
     * `parameters.csv`: the parameter file, created by `raz_create_parameter_files`
     * `mbd.tree`: the true MBD tree, created by `raz_create_input_files`
     * `mbd.fasta`: the true MBD alignment, created by `raz_create_input_files`
     * `bd.tree`: the win BD tree, created by `raz_create_input_files`
     * `bd.fasta`: the twin BD alignment, created by `raz_create_input_files`
     * `mbd.trees`: the posterior trees from `mbd.tree`, created by `raz_create_inference_files("mbd.tree")`
     * `mbd.log`: the posterior parameter estimates from `mbd.tree`, created by `raz_create_inference_files("mbd.tree")`
     * `mbd_mar_lik.csv`: the posterior's marginal likelihood from `mbd.tree`, created by `raz_create_inference_files("mbd.tree")`
     * `bd.trees`: the posterior trees from `bd.tree`, created by `raz_create_inference_files("bd.tree")`
     * `bd.log`: the posterior parameter estimates from `bd.tree`, created by `raz_create_inference_files("bd.tree")`
     * `bd_mar_lik.csv`: the posterior's marginal likelihood from `bd.tree`, created by `raz_create_inference_files("bd.tree")`
     * `mbd_nltts.csv`: the nLTT statistic distribution between `mbd.tree` and `mbd.trees`, created by `raz_create_nltt_file("mbd.tree", "mbd.trees")`
     * `bd_nltts.csv`: the nLTT statistic distribution between `bd.tree` and `bd.trees`, created by `raz_create_nltt_file("bd.tree", "bd.trees")`
  * `2`
  * `3`
  * Etcetera
     
## Overview

![Pipeline](pipeline.png)

Load the library:

```{r}
library(razzo)
```

We will work in this folder:

```{r}
folder_name <- tempdir()
```

Here we create all parameter files:

```{r}
parameters_filenames <- raz_create_parameters_files(folder_name)
testit::assert(file.path(folder_name, "1", "parameters.csv") %in% parameters_filenames)
```
Each parameter file is read to create a MBD tree and alignment and its BD twin:

```{r}
# Create all true trees, true alignments and their twins
for (parameters_filename in parameters_filenames) {
  input_filenames <- raz_create_input_files(parameters_filename)
  # True MBD tree
  testit::assert(file.path(folder_name, "1", "mbd.tree") %in% input_filenames)
  # True MBD alignment
  testit::assert(file.path(folder_name, "1", "mbd.fasta") %in% input_filenames)
  # Twin BD tree
  testit::assert(file.path(folder_name, "1", "bd.tree") %in% input_filenames)
  # Twin BD alignment
  testit::assert(file.path(folder_name, "1", "bd.fasta") %in% input_filenames)
}

# Do the inference
fasta_filenames <- c("1a.fasta") # Search the folder
for (fasta_filename in fasta_filenames) 
{
  output_filenames <- raz_create_inference_files(fasta_filename)
  # Posterior trees
  testit::assert("1a.trees" %in% output_filenames)
  # Trace of MCMC, to estimate the Effective Sample Sizes
  testit::assert("1a.log" %in% output_filenames)
  # Marginal likelihood
  testit::assert("1a_mar_lik.csv" %in% output_filenames)
}

# Create the nLTT distribution
trees_filenames <- c("1a.trees") # Search the folder
for (trees_filename in trees_filenames)
{
  nltt_filename <- raz_create_nltt_file(trees_filename)
  testit::assert("1a_nltts.csv" %in% nltt_filename)
}

# All files are in place!
```

### Step 0: create parameters


### step 1: simulate MBD tree

Set seed:

```{r}
seed <- 1
```

First we want to simulate an MBD tree.

```{r}
MBD_pars <- c(0.2, 0.15, 2, 0.10)
soc <- 2
age <- 10
cond <- 1
set.seed(seed)
MBD_sim  <- mbd::mbd_sim(pars = MBD_pars, soc = soc, age = age, cond = cond)
MBD_brts <- MBD_sim$brts
MBD_tree <- MBD_sim$tes
MBD_l_matrix <- MBD_sim$l_matrix
plot(MBD_tree)
```

### step 2: calculate the best pars for the twin BD tree (be careful with the estimated parameters!)
Then we wanto to generate a "twin" BD tree. To create such tree we have to estimate the best lambda and mu to make the comparison fair. 
To estimate the parameters we run a ML inference using the BD model.
We also want to condition on the same amount of tips and on the survival of the phylogeny (cond = 2).

```{r}
BD_pars <- DDD::bd_ML(brts = abs(MBD_sim$brts),
                      cond = 2,
                      initparsopt = MBD_pars[1:2],
                      idparsopt = 1:2,
                      missnumspec = 0,
                      tdmodel = 0,
                      btorph = 1,
                      soc = soc); BD_pars

# BD_pars2 <- DDD::bd_ML(brts = abs(MBDsim$brts),
#                       cond = 1,
#                       initparsopt = simpars[1:2],
#                       idparsopt = 1:2,
#                       missnumspec = 0,
#                       tdmodel = 0,
#                       btorph = 1,
#                       soc = soc)

```


### step 3: simulate the twin BD tree
###(check if the method to keep the topology is right. from the figures it seems ok-ish. check with small tree)
Now that we have the BD equivalent parameters for a fair "twinning" we can simulate a BD tree.
We use "TESS::tess.sim.taxa.age" to generate the branching times (BD_brts). 
We then plug these branching times in the l_table coming from the original MBD tree to be sure that the topology is the same.

```{r simulate_twin_bd_tree}
if (1 == 2) {
  set.seed(seed)
  BD_tree0 <- TESS::tess.sim.taxa.age(n = 1,
                                      lambda = as.numeric(unname(BD_pars[1])),
                                      mu     = as.numeric(unname(BD_pars[2])),
                                      nTaxa = ((soc - 1) + length(MBD_sim$brts)),
                                      age = age,
                                      MRCA = TRUE)[[1]]
  
  BD_brts <- ape::branching.times(BD_tree0)
  BD_l_matrix <- MBD_l_matrix
  alive <- BD_l_matrix[, 4] == -1
  alive2 <- alive
  t <- length(alive); while (sum(alive2) > length(BD_brts)) {if (alive2[t] == 1) {alive2[t] = 0}; t = t - 1}
  vec <- BD_l_matrix[, 1]
  vec[seq_along(vec) * alive2] <- BD_brts
  BD_l_matrix[, 1] <- vec
  par(mfrow = c(1,2))
  BD_tree <- DDD:::L2phylo(BD_l_matrix)
  plot(MBD_tree, main = "MBD tree"); plot(BD_tree, main = "twin BD tree")
}
```

###step 4: generate MBD posterior, given BD prior
The next step is to use the package "pirouette" to generate, from the MBD tree, first the alignments and then, through BEAST, a posterior distribution of trees. These trees are obtained using a BD prior. This means that there are no simultaneous branching events.

```{r}
if (1 == 2) {
  MBD_sequence_length <- 1e3;
  MBD_mutation_rate   <- 1e-3;
  MBD_chain_length    <- 1e6;
  MBD_sample_interval <- 1e3;
  set.seed(seed)
  MBD_posterior <- pirouette::pir_run( #test1 is a posterior of 10001 trees generated from sim_tes[[1]]
    phylogeny = MBD_tree,
    sequence_length = MBD_sequence_length,
    mutation_rate = MBD_mutation_rate,
    mcmc = beautier::create_mcmc(chain_length = MBD_chain_length, store_every = MBD_sample_interval), #store_every = -1
    site_models = beautier::create_jc69_site_model(),
    clock_models = beautier::create_strict_clock_model(),
    tree_priors = beautier::create_bd_tree_prior(),
    mrca_distr = beautier::create_normal_distr(mean  = beautier::create_mean_param(value = age), 
                                               sigma = beautier::create_sigma_param(value = 0.001)),
    alignment_rng_seed = 0,
    beast2_rng_seed = 1,
    verbose = FALSE
  )
}
```

###step 5: generate BD posterior, given BD prior
###(check BD_substitution rate, see Rampal's email)
We repeat the same process for the tree simulated according to the BD model. We want to have, in principle, the same amount of substitutions; to do so we modify the mutation rate according to the ratio of the total branch lenghts of the trees simulated with the two models.

```{r}
if (1 == 2) {
  BD_sequence_length <- MBD_sequence_length;
  BD_mutation_rate   <- MBD_mutation_rate * (sum(MBD_tree$edge.length)/sum(BD_tree$edge.length));
  BD_chain_length    <- MBD_chain_length;
  BD_sample_interval <- MBD_sample_interval;
  set.seed(seed)
  BD_posterior <- pirouette::pir_run( #test1 is a posterior of 10001 trees generated from sim_tes[[1]]
    phylogeny = BD_tree,
    sequence_length = BD_sequence_length,
    mutation_rate = BD_mutation_rate,
    mcmc = beautier::create_mcmc(chain_length = BD_chain_length, store_every = BD_sample_interval), #store_every = -1
    site_models = beautier::create_jc69_site_model(),
    clock_models = beautier::create_strict_clock_model(),
    tree_priors = beautier::create_bd_tree_prior(),
    mrca_distr = beautier::create_normal_distr(mean = beautier::create_mean_param(value = age), sigma = beautier::create_sigma_param(value = 0.001)),
    alignment_rng_seed = 0,
    beast2_rng_seed = 1,
    verbose = FALSE
  )
}
```

### step 6: calculating nLTT statistics for MBD and BD
Finally we calculate the nLTT statistics for both posteriors related to the original trees.

```{r}
if (1 == 2) {
  MBD_nLTT.diff <- rep(NA, length(MBD_posterior$trees))
  for (i in 1:length(MBD_posterior$trees))
  {
    MBD_nLTT.diff[i] <- nLTT::nLTTstat(MBD_tree, MBD_posterior$trees[[i]])
  }
  MBD_mean.nLTT <- mean(MBD_nLTT.diff)
  MBD_std.nLTT  <- sqrt(stats::var(MBD_nLTT.diff)); MBD_std.nLTT
  MBD_df.nLTT   <- data.frame(diff = MBD_nLTT.diff)
  
  BD_nLTT.diff <- rep(NA, length(BD_posterior$trees))
  for (i in 1:length(BD_posterior$trees))
  {
    BD_nLTT.diff[i] <- nLTT::nLTTstat(BD_tree, BD_posterior$trees[[i]])
  }
  BD_mean.nLTT <- mean(BD_nLTT.diff)
  BD_std.nLTT  <- sqrt(stats::var(BD_nLTT.diff)); BD_std.nLTT
  BD_df.nLTT   <- data.frame(diff = BD_nLTT.diff)
  
  par(mfrow = c(1,2))
  hist(unlist(MBD_df.nLTT), main = "MBD nLTT")
  hist(unlist(BD_df.nLTT), main = "BD nLTT")
  cat("Average nLTT for MBD", MBD_mean.nLTT, "\nAverage nLTT for BD ", BD_mean.nLTT)
}
```


