---
title: "razzo pipeline"
author: "G. Laudanno and Richel J.C. Bilderbeek"
date: "`r Sys.Date()`"
output: rmarkdown::pdf_document
vignette: >
  %\VignetteIndexEntry{razzo pipeline}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

Legenda:
nLTT = normalized lineages through time;
BD = Birth Death Model;
MBD = Multiple Birth Death Model;

This is the pipeline to follow for the Razzo project. Our aim is to compare two nLTT distributions related to two tree posteriors: one descending from MBD alignments and another one from BD alignments. Both posteriors are generated using a BD prior in BEAST.

## Folder structure

In `razzo_project`, there is a folder called `data`.
In `data`, there are folders named after their parameters, e.g. `0.2-0.15-1-0.1`.
In each of these folders, there are folders named after their seeds, e.g. `1`.
In each of these folders, there are:

Filename|Description|Created by
-------------|---------------------------------------|--------------------------
`parameters.csv`|the parameter file|`raz_create_parameter_files`
`mbd.tree`|the true MBD tree|`raz_create_mbd_tree_file`
`mbd.fasta`|the true MBD alignment|`raz_create_mbd_alignment_file`
`bd.tree`|the twin BD tree|`raz_create_bd_tree_file`
`bd.fasta`|the twin BD alignment|`raz_create_bd_alignment_file`
`mbd.trees`|the posterior trees from `mbd.tree`|`raz_create_mbd_posterior_files`
`mbd.log`|the posterior parameter estimates from `mbd.tree`|`raz_create_mbd_posterior_files`
`mbd_mar_lik.csv`|the posterior's marginal likelihood from `mbd.tree`|`raz_create_mbd_posterior_files`
`bd.trees`|the posterior trees from `bd.tree`|`raz_create_bd_posterior_files`
`bd.log`|the posterior parameter estimates from `bd.tree`|`raz_create_bd_posterior_files`
`bd_mar_lik.csv`|the posterior's marginal likelihood from `bd.tree`|`raz_create_mb_posterior_files`
`mbd_nltts.csv`|the nLTT statistic distribution between `mbd.tree` and `mbd.trees`|`raz_create_mbd_nltt_file`
`bd_nltts.csv`|the nLTT statistic distribution between `bd.tree` and `bd.trees`|`raz_create_bd_nltt_file`

## Overview

![Pipeline](pipeline.png)

Update our dependencies:

```{r message=FALSE}
if (1 == 2) {
  devtools::install_github("Giappo/mbd", quiet = TRUE)
  devtools::install_github("richelbilderbeek/beautier", quiet = TRUE)
  devtools::install_github("richelbilderbeek/tracerer", quiet = TRUE)
  devtools::install_github("richelbilderbeek/beastier", quiet = TRUE)
  devtools::install_github("richelbilderbeek/mauricer", quiet = TRUE)
  devtools::install_github("richelbilderbeek/babette", quiet = TRUE)
}
```

Load the library:

```{r}
library(razzo)
```

We will work in this folder:

```{r}
super_folder_name <- tempdir()
project_folder_name <- file.path(super_folder_name, "razzo_project") 
dir.create(path = project_folder_name, recursive = TRUE)
```

## Step 1: Create parameter files

Here we create all parameter files:

```{r}
parameters_filenames <- raz_create_parameters_files(
  project_folder_name = project_folder_name
)
```

To speed things up, we'll be keeping only two parameter files:

```{r}
parameters_filenames <- parameters_filenames[ c(1, 2) ]
knitr::kable(parameters_filenames)
```

## Step 2: create MBD trees

```{r}
mbd_tree_filenames <- parameters_filenames
# Create all true trees, true alignments and their twins
for (i in seq_along(parameters_filenames)) {
  parameters_filename <- parameters_filenames[i]
  mbd_tree_filenames[i] <- raz_create_mbd_tree_file(parameters_filename)
}
```

Taking a look at the first MBD tree:

```{r}
ape::plot.phylo(ape::read.tree(file = mbd_tree_filenames[1]))
```

## Step 3: create twin BD trees

Then we want to generate a "twin" BD tree. To create such tree we have to estimate the best lambda and mu to make the comparison fair. 
To estimate the parameters we run a ML inference using the BD model.
We also want to condition on the same amount of tips and on the survival of the phylogeny (cond = 2).

```{r}
bd_tree_filenames <- parameters_filenames
# Create all BD twin trees, true alignments and their twins
for (i in seq_along(parameters_filenames)) {
  parameters_filename <- parameters_filenames[i]
  bd_tree_filenames[i] <- raz_create_bd_tree_file(parameters_filename)
}
```

Taking a look at the first MBD tree:

```{r}
ape::plot.phylo(ape::read.tree(file = bd_tree_filenames[1]))
```

## Step 4: Create MBD alignments

```{r}
mbd_alignment_filenames <- parameters_filenames
# Create all true trees, true alignments and their twins
for (i in seq_along(parameters_filenames)) {
  parameters_filename <- parameters_filenames[i]
  mbd_alignment_filenames[i] <- raz_create_mbd_alignment_file(parameters_filename)
}
```

Taking a look at the first MBD alignment:

```{r}
image(ape::read.FASTA(file = mbd_alignment_filenames[1]))
```

## Step 5: Create BD alignments

```{r}
bd_alignment_filenames <- parameters_filenames
# Create all true trees, true alignments and their twins
i <- 1
for (i in seq_along(parameters_filenames)) {
  parameters_filename <- parameters_filenames[i]
  bd_alignment_filenames[i] <- raz_create_bd_alignment_file(parameters_filename)
}
```

Taking a look at the first MBD alignment:

```{r}
image(ape::read.FASTA(file = bd_alignment_filenames[1]))
```

# UP TILL HERE

## Step 2: create inference files

```{r}
if (1 == 2) {
  # Do the inference. It doesn't work on Windows.
  mbd_alignment_filenames <- list() # Search the folder
  for (i in seq_along(parameters_filenames)) {
    parameter_filename <- parameters_filenames[i] 
    mbd_alignment_filenames[[i]] <- raz_create_mbd_posterior_files(
      parameter_filename
    )
    # Posterior trees
    testit::assert("1a.trees" %in% output_filenames)
    # Trace of MCMC, to estimate the Effective Sample Sizes
    testit::assert("1a.log" %in% output_filenames)
    # Marginal likelihood
    testit::assert("1a_mar_lik.csv" %in% output_filenames)
  }
}
```

## step 6: calculating nLTT statistics for MBD and BD

Finally we calculate the nLTT statistics for both posteriors related to the original trees.

```{r}
if (1 == 2) {
  # Create the nLTT distribution
  trees_filenames <- c("1a.trees") # Search the folder
  for (trees_filename in trees_filenames)
  {
    if (1 == 2) {
      # TODO: Issue 5
      nltt_filename <- raz_create_nltt_file(trees_filename)
      testit::assert("1a_nltts.csv" %in% nltt_filename)
    }
  }
}
# All files are in place!
```

```{r}
# TODO: plot the nLTT
```

### step 3: simulate the twin BD tree
###(check if the method to keep the topology is right. from the figures it seems ok-ish. check with small tree)
Now that we have the BD equivalent parameters for a fair "twinning" we can simulate a BD tree.
We use "TESS::tess.sim.taxa.age" to generate the branching times (BD_brts). 
We then plug these branching times in the l_table coming from the original MBD tree to be sure that the topology is the same.

###step 4: generate MBD posterior, given BD prior
The next step is to use the package "pirouette" to generate, from the MBD tree, first the alignments and then, through BEAST, a posterior distribution of trees. These trees are obtained using a BD prior. This means that there are no simultaneous branching events.

###step 5: generate BD posterior, given BD prior
###(check BD_substitution rate, see Rampal's email)
We repeat the same process for the tree simulated according to the BD model. We want to have, in principle, the same amount of substitutions; to do so we modify the mutation rate according to the ratio of the total branch lenghts of the trees simulated with the two models.


```{r}
if (1 == 2) {
  
  graphics::par(mfrow = c(1, 2))
  hist(unlist(MBD_df.nLTT), main = "MBD nLTT")
  hist(unlist(BD_df.nLTT), main = "BD nLTT")
  cat("Average nLTT for MBD", MBD_mean.nLTT, "\nAverage nLTT for BD ", BD_mean.nLTT)
}
```

